src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
+                           neighbor_neighbors_list[j]);
     }
 
-    idType next_closest_entry_point = selectedNeighbors.back();
-    {
-        idType *ll_cur = get_linklist_at_level(cur_c, level);
-        assert(getListCount(ll_cur) == 0 &&
-               "The newly inserted element should have blank link list");
-        const linkListSize size = selectedNeighbors.size();
-        setListCount(ll_cur, size);
+    std::vector<idType> nodes_to_update;
+    auto orig_candidates = candidates;
 
-        for (auto cur_neighbor = selectedNeighbors.rbegin();
-             cur_neighbor != selectedNeighbors.rend(); ++cur_neighbor) {
+    // Candidates will store the newly selected neighbours (for the neighbor).
+    size_t max_M_cur = level ? maxM_ : maxM0_;
+    getNeighborsByHeuristic2(candidates, max_M_cur);
 
-            assert(*ll_cur == 0 && "Possible memory corruption");
-            assert(level <= element_levels_[*cur_neighbor] &&
-                   "Trying to make a link on a non-existent level");
+    // Go over the original candidates set, and save the ones chosen to be removed to update later
+    // on.
+    bool cur_node_chosen = false;
+    while (orig_candidates.size() > 0) {
+        idType orig_candidate = orig_candidates.top().second;
+        if (candidates.empty() || orig_candidate != candidates.top().second) {
+            if (orig_candidate != new_node_id) {
+                nodes_to_update.push_back(orig_candidate);
+            }
+            orig_candidates.pop();
+        } else {
+            candidates.pop();
+            orig_candidates.pop();
+            if (orig_candidate == new_node_id) {
+                cur_node_chosen = true;
+            }
+        }
+    }
+
+    // Acquire all relevant locks for making the updates for the selected neighbor - all its removed
+    // neighbors, along with the neighbors itself and the cur node.
+    // but first, we release the node and neighbors lock to avoid deadlocks.
+    node_lock.unlock();
+    neighbor_lock.unlock();
+
+    nodes_to_update.push_back(selected_neighbor);
+    nodes_to_update.push_back(new_node_id);
+
+    std::sort(nodes_to_update.begin(), nodes_to_update.end());
+    size_t nodes_to_update_count = nodes_to_update.size();
+    std::unique_lock<std::mutex> locks[nodes_to_update_count];
+    for (size_t i = 0; i < nodes_to_update_count; i++) {
+        locks[i] = std::unique_lock<std::mutex>(element_neighbors_locks_[nodes_to_update[i]]);
+    }
 
-            *ll_cur = *cur_neighbor;
-            ll_cur++;
+    auto *neighbour_incoming_edges = getIncomingEdgesPtr(selected_neighbor, level);
+    size_t neighbor_neighbors_count = getListCount(neighbor_neighbors_list);
+
+    size_t neighbour_neighbours_idx = 0;
+    bool update_cur_node_required = true;
+    for (size_t i = 0; i < neighbor_neighbors_count; i++) {
+        if (!std::binary_search(nodes_to_update.begin(), nodes_to_update.end(),
+                                neighbor_neighbors_list[i])) {
```

Comment:
is this tested to be faster than a regular search?

Reply:
What do you mean by regular search? Since we know that `nodes_to_update` is already sorted, using binary search is the trivially better option no?